Household composition repeat single

samiwel opened the pull request:
### What is the context of this PR?
Added a new question type for repeating answers.
Created a new schema to demonstrate the household composition question.
Implement repeated single answers using the new Answer Store.

### How to review 
Should be able to add and remove people on the household question screen.
The answers should be persisted in the Answer Store.
N.B. This version does not make use of Javascript and only deals with single answers repeating. Multiple answers and introducing Javascript are covered by different stories.

samiwel committed:
Add repeating answer question type.

samiwel committed:
Split action into separate endpoint.

samiwel committed:
Add separate remove endpoint.

samiwel committed:
Add answer store for structured storage of flattened answers.

  - Replace question store answers list with new type AnswerStore.
  - Fix basic loading and saving of answer store data.

samiwel committed:
Fix issue storing flattened answers.

samiwel committed:
WIP answer store changes.

samiwel committed:
WIP Fix issue adding new people.

samiwel committed:
WIP Convert answer store back to dict.

samiwel committed:
WIP Fix answer store not serialising answers correctly.

samiwel committed:
WIP Fix state construction issue.

iwootten committed:
Apply answer store changes.

hamishtaplin committed:
styled household question

hamishtaplin committed:
Merge branch 'household-composition-repeat-single' of github.com:ONSdigital/eq-survey-runner into household-composition-repeat-single

iwootten committed:
Apply changes to questionnaire manager.

iwootten committed:
Add clear method to empty answer store

iwootten committed:
Refactor method name

iwootten committed:
Move to using instance method for answer map

hamishtaplin committed:
added basic add/remove mechanism for people in household

samiwel committed:
Fix remove link and summary display..

samiwel committed:
Merge branch 'household-composition-repeat-single' of github.com:ONSdigital/eq-survey-runner into household-composition-repeat-single

hamishtaplin committed:
implemented renumbering of household items when an item/s is removed

hamishtaplin committed:
disabled JS implementation of household question

samiwel committed:
Fix code style issues.

samiwel committed:
Merge branch 'household-composition-repeat-single' of github.com:ONSdigital/eq-survey-runner into household-composition-repeat-single

samiwel committed:
Fix failing unit test.

iwootten committed:
Return an order dict from map

samiwel committed:
Change suffix of answers to use underscores.

samiwel committed:
Fix ordering on summary screen.

samiwel committed:
Fix failing selenium test.

samiwel committed:
Refactor code.

samiwel committed:
Add some unit tests.

samiwel committed:
Merge branch 'master' into household-composition-repeat-single

samiwel committed:
Refactor code following merge.

jonnyshaw89 commented:
Think you committed your code coverage report?

samiwel committed:
Remove coverage report from git.

samiwel committed:
Add html coverage report to .gitignore.

samiwel committed:
Fix quantified code issue.

collisdigital commented on app/templates/partials/question.html:
```diff
@@ -24,13 +24,58 @@ <h3 class="question__subtitle mars">{{question_subtitle}}</h3>
+              form_type=meta.survey.form_type,
+              location=content.id,
+              question=question.schema_item.id) }}"
+            name="action[remove_answer]" value="{{ loop.index0 }}">Remove <span class="u-vh">Person {{loop.index0}}</span></button>
```
These kinds of strings "Remove" and "Person" need to translatable to Welsh.

collisdigital commented on app/templates/partials/question.html:
```diff
@@ -24,13 +24,58 @@ <h3 class="question__subtitle mars">{{question_subtitle}}</h3>
+            form_type=meta.survey.form_type,
+            location=content.id,
+            question=question.schema_item.id) }}"
+          name="action[add_answer]">Add another person</button>
```
Same for translating this string

samiwel committed:
Add Welsh translation support for template strings.

samiwel committed:
WIP Adding selenium tests.

samiwel committed:
Merge branch 'master' into household-composition-repeat-single

samiwel committed:
Refactor state building for repeating answers.

samiwel committed:
Refactor code and Add unit tests for household composition question.

  - Fix codacy issues.
  - Fix quantified code issues.
  - Add unit tests for repeating answers.

benjefferies commented on app/data/test_household_question.json:
```diff
@@ -0,0 +1,74 @@
+                                  "answers": [
+                                    {
+                                        "display": {},
+                                        "id": "414699da-1667-44fd-8e98-7606966884db",
```
Is it worth calling this household_full_name

samiwel replied:
Yes, probably if we're deciding to no longer use GUIDs.

ajmaddaford replied:
Yes, but use hyphens rather than underscores

hamishtaplin commented on app/templates/partials/question.html:
```diff
@@ -24,13 +24,58 @@ <h3 class="question__subtitle mars">{{question_subtitle}}</h3>
+            <span class="js-household-action">{{ _('You') }}</span>
+         {%- else -%}
+         <button class="btn btn--link"
+            formaction="{{ url_for('action.remove_answer',
```
We can't rely on `formaction` as it's a HTML5 attribute and so won't work in older browsers.

hamishtaplin committed:
cleaned up question template a bit

jonnyshaw89 commented on app/assets/js/app/main.js:
```diff
@@ -4,3 +4,4 @@ import './modules/accordion'
 import './modules/analytics'
 import './modules/inpagelink'
 import './modules/focus-styles'
+// import './modules/household'
```
Remove this comment

samiwel replied:
@jonnyshaw89 This needs to remain commented out for now. It should be uncommented when we work on **Household Composition Entry: Realtime Changes** story.

hamishtaplin commented on app/templates/partials/answer.html:
```diff
@@ -1,7 +1,8 @@
-  {% set invalid = answer.is_valid == False %}
+{% set invalid = answer.is_valid == False %}
+
+<div class="answer answer--{{answer.schema_item.type|lower}} {{'js-has-errors' if invalid}}" id="{{answer.id}}">
```
The `answer.id` here is not unique when we add new answers. This needs to be fixed at a high level (not in the template) as that id is then used to generate id's further down the tree.

samiwel replied:
I've modified the backend code to return unique Id's for each instance of an answer when it renders the widget.

hamishtaplin reviewed (changes requested):
A couple of issues you may already know about:
1. When you add multiple people and then submit without entering a name, the error at the top of the page doesn't reflect this (only one error).
2. When you submit with errors and then add another person, the errors aren't maintained.

benjefferies commented on app/assets/js/app/main.js:
```diff
@@ -4,3 +4,4 @@ import './modules/accordion'
 import './modules/analytics'
 import './modules/inpagelink'
 import './modules/focus-styles'
+// import './modules/household'
```
Can this be deleted?

samiwel replied:
See https://github.com/ONSdigital/eq-survey-runner/pull/563#discussion_r88448316

hamishtaplin commented on app/templates/partials/question.html:
```diff
@@ -24,13 +24,55 @@ <h3 class="question__subtitle mars">{{question_subtitle}}</h3>
+              <span class="js-household-action">{{ _('You') }}</span>
+           {%- else -%}
+           <button class="btn btn--link"
+              formaction="{{ url_for('action.remove_answer',
```
`formaction` is a html5 attribute and can't be relied on.

benjefferies replied:
Not sure how I feel able this, maybe we should just hard code the url with the dynamic parts substituted or create a add/remove variable in the context

jonnyshaw89 commented on app/main/views/questionnaire.py:
```diff
@@ -119,13 +125,31 @@ def submit_answers(eq_id, form_type, collection_id):
+@action_blueprint.route('<location>/<question>/add', methods=["POST"])
+@login_required
+def add_answer(eq_id, form_type, collection_id, location, question):
+    answer_store = get_answer_store(current_user)
```
we are just loading answer store from global to pass it into questionnaire_manager. should we just load it in questionnaire_manager?

samiwel replied:
The reason for passing it in is that it is easier to use a mock answer store during unit tests. If you can show me how to mock the global call to get_answer_store within questionnaire_manager.add_answer, then I'm happy to move it.

benjefferies commented on tests/app/data_model/test_answer_store.py:
```diff
@@ -267,3 +267,93 @@ def test_returns_ordered_map(self):
+            "2_1": 25,
+            "5_1": 65,
+        }
+        self.assertEqual(self.store.map(), expected_answers)
```
Do we need this assertion, does that mean this test is testing the add and remove functionality?

samiwel replied:
I split these out into two tests.

benjefferies commented on app/main/views/questionnaire.py:
```diff
@@ -30,8 +30,13 @@
                                     import_name=__name__,
                                     url_prefix='/questionnaire/<eq_id>/<form_type>/<collection_id>/')
 
+action_blueprint = Blueprint(name='action',
```
Would it be cleaner if we have the action_blueprint in another file?

samiwel replied:
Not sure if we need the blueprint at all. There's been some discussion about having a single endpoint to avoid using HTML5 formaction attribute in the template. If I re-work this so that it posts back to the same endpoint then we may just be able to use the questionnaire endpoint or some variation on it.

jonnyshaw89 commented on app/questionnaire/questionnaire_manager.py:
```diff
@@ -125,3 +126,44 @@ def get_schema_item_by_id(self, item_id):
+        if self.state is None:
+            self.process_incoming_answers(block, post_data)
+
+        index_to_remove = post_data.get('action[remove_answer]')
```
maybe we can pass the index through to this function? IMO references to the name of a button should probably by up at the 1st level rather than being extracted lower down.

samiwel replied:
Done.

jonnyshaw89 commented on app/questionnaire_state/state_item.py:
```diff
@@ -38,3 +38,10 @@ def get_errors(self):
+            if child.schema_item == schema_item:
+                return child
+            else:
+                return child.find_state_item(schema_item)
```
what if none of the children contain the schema_item?

benjefferies replied:
Python has an implicit none return but should we be explicit there is a valid question

samiwel replied:
I've made this more explicit so that it returns None if schema item is not found.

samiwel committed:
Fix error during submission of repeating answers.

benjefferies commented on app/questionnaire/questionnaire_manager.py:
```diff
@@ -125,3 +126,44 @@ def get_schema_item_by_id(self, item_id):
         return self._schema
+
+    def add_answer(self, block, question, post_data, answer_store):
+        if self.state is None:
```
Should we call process_incoming_answers in questionnaire.py as that would be consistent with what we are currently doing

samiwel replied:
Moved this logic into questionnaire.py.

samiwel committed:
Merge branch 'household-composition-repeat-single' of github.com:ONSdigital/eq-survey-runner into household-composition-repeat-single

benjefferies commented on app/questionnaire/questionnaire_manager.py:
```diff
@@ -125,3 +126,44 @@ def get_schema_item_by_id(self, item_id):
+        return new_answer
+
+    def remove_answer(self, block, post_data, answer_store):
+        if self.state is None:
```
Should we call process_incoming_answers in questionnaire.py as that would be consistent with what we are currently doing

jonnyshaw89 commented on app/questionnaire_state/state_question.py:
```diff
@@ -6,3 +8,54 @@ def __init__(self, id, schema_item):
+
+            for child in state_answers:
+                answer_keys = [key if key.startswith(child.schema_item.id) else None for key in user_input]
+                answer_instances = sorted(list(filter(None, answer_keys)))
```
What is this line doing? filtering by nothing?

samiwel replied:
Filters out answers starting with the answer Id from the schema. The filter on None discards values in user_input that are not related to the current child.

jonnyshaw89 commented on app/templates/partials/answer.html:
```diff
@@ -1,7 +1,8 @@
-  {% set invalid = answer.is_valid == False %}
+{% set invalid = answer.is_valid == False %}
+
+<div class="answer answer--{{answer.schema_item.type|lower}} {{'js-has-errors' if invalid}}" id="{{answer.id}}">
```
Instead of `if invalid` can we not do something like `if not answer.is_valid`?
Im not sure if thats possible

hamishtaplin replied:
@jonnyshaw89 That `invalid` value is used in two places so this caches the result so the logic isn't repeated.

benjefferies commented on app/questionnaire_state/state_question.py:
```diff
@@ -6,3 +8,54 @@ def __init__(self, id, schema_item):
+    def update_state(self, user_input):
+
+        if self.schema_item.type == 'RepeatingAnswer':
+            state_answers = self.children[:]
```
Maybe this refactored into a method with a name like build_repeating_answers or something would give that extra bit of context as there is quite a bit of logic here

samiwel replied:
Done.

benjefferies commented on app/questionnaire_state/state_question.py:
```diff
@@ -6,3 +8,54 @@ def __init__(self, id, schema_item):
+                answer_instances = sorted(list(filter(None, answer_keys)))
+                num_instances = len(answer_instances)
+
+                if num_instances > 1:
```
Do we need `if num_instances > 1:`? Wouldn't it just iterate once and call child.update_state

samiwel replied:
Yeah, I think you're right. I'll remove it.

samiwel replied:
Actually, I still think we need this. Imagine what happens in the case of {} being passed in.

benjefferies replied:
Would num_instances not be 0 then and it wouldn't iterate anyway?

samiwel replied:
I'll give it a test it might be okay to remove it after all. Confusing myself because it's changed to build the state on update_state now rather than construct state.

benjefferies commented on app/submitter/converter.py:
```diff
@@ -88,7 +88,13 @@ def prepare_answers(metadata_store, questionnaire, answers):
                 value = answers[key]
                 if value is not None:
-                    data[item.code] = Converter._encode_value(value)
+                    if item.code not in data:
```
After a quick discussion with @ajmaddaford it might makes sense to not change anything here until we know the format we send this to the downstream system in

samiwel replied:
Had to do something or you get an error when you try to submit. Happy to have a discussion about what the downstream format should/needs to be.

ajmaddaford replied:
This is fine for now, we need to refactor the submitter to use the answer store, but that can be done in another pull request.

ajmaddaford commented on app/questionnaire/questionnaire_manager.py:
```diff
@@ -125,3 +126,44 @@ def get_schema_item_by_id(self, item_id):
     def get_schema(self):
         return self._schema
+
+    def add_answer(self, block, question, post_data, answer_store):
```
It would be clearer to rename `block` to `block_id` and `question` to `question_id` to make it clear they are id's and not objects

samiwel replied:
Changed.

ajmaddaford commented on app/questionnaire/questionnaire_manager.py:
```diff
@@ -125,3 +126,44 @@ def get_schema_item_by_id(self, item_id):
+        question_schema = self._schema.get_item_by_id(question)
+        question_state = self.state.find_state_item(question_schema)
+
+        answer_schema = question_schema.answers[0]  # Single answer for now.
```
I think renaming `answer_schema` to `existing_answer_schema` would make things clearer

samiwel replied:
Changed

ajmaddaford commented on app/questionnaire/questionnaire_manager.py:
```diff
@@ -125,3 +126,44 @@ def get_schema_item_by_id(self, item_id):
+        question_state = self.state.find_state_item(question_schema)
+
+        answer_schema = question_schema.answers[0]  # Single answer for now.
+        new_answer = self._create_new_answer(answer_schema, question_state, answer_store)
```
I think renaming `new_answer` to `new_answer_state` and `_ create_new_answer` to `_create_new_answer_state` would make it clear we are creating a new state item here.

samiwel replied:
Changed

benjefferies commented on app/templates/partials/question.html:
```diff
@@ -24,13 +24,55 @@ <h3 class="question__subtitle mars">{{question_subtitle}}</h3>
+  {%- if question.schema_item.type == 'RepeatingAnswer' %}
+  <div class="question__actions">
+    <button class="btn btn--border btn--secondary venus js-household-btn" type="submit"
+          formaction="{{ url_for('action.add_answer',
```
Not sure how I feel able this, maybe we should just hard code the url with the dynamic parts substituted or create a add/remove variable in the context

ajmaddaford commented on app/questionnaire/questionnaire_manager.py:
```diff
@@ -125,3 +126,44 @@ def get_schema_item_by_id(self, item_id):
+        new_answer = self._create_new_answer(answer_schema, question_state, answer_store)
+
+        question_answers = question_state.children
+        question_answers.append(new_answer)
```
Can we just do `question_state.answers.append(new_answer_state)`?

samiwel replied:
Changed

benjefferies commented on tests/app/data_model/test_answer_store.py:
```diff
@@ -267,3 +267,93 @@ def test_returns_ordered_map(self):
+        }
+        self.assertEqual(self.store.map(), expected_answers)
+
+    def test_remove_answer_(self):
```
This is the same as above?

samiwel replied:
The first assert is probably superfluous as I think there is already a separate test for the add. All I'm doing here is asserting that we did have 2 answers and after the remove we only have one.

samiwel replied:
Removed duplicate test.

samiwel committed:
Merge branch 'master' into household-composition-repeat-single

benjefferies commented on tests/app/questionnaire/test_questionnaire_manager.py:
```diff
@@ -0,0 +1,145 @@
+        self.assertEqual(new_answer.schema_item.widget.name, 'answer_2')
+        self.assertEqual(new_answer.instance, 2)
+
+    def setUp(self):
```
Minor point but it's a bit confusing reading the tests with this at the bottom as you have references to things you don't expect to exist but that might be a preference thing

benjefferies commented on tests/app/questionnaire_state/test_state_item.py:
```diff
@@ -0,0 +1,44 @@
+        # Then
+        self.assertEqual(result, child)
+
+    def test_find_schema_item_recursive_call_if_not_found(self):
```
Test names might be nicer test_find_schema_item_in_question and test_find_schema_item_in_block as I think that's what it's testing right? Again minor (possibly preference)

samiwel replied:
Agreed, changed.

benjefferies commented on tests/app/questionnaire_state/test_state_question.py:
```diff
@@ -0,0 +1,152 @@
+
+        question.answers.append(answer)
+
+        self.assertTrue(answer in question.answers)
```
Should these assertions be another test?

samiwel replied:
I've split these out into two separate tests.

benjefferies commented on tests/app/questionnaire_state/test_state_question.py:
```diff
@@ -0,0 +1,152 @@
+        question.answers.append(answer1)
+        question.answers.append(answer2)
+
+        self.assertEqual(len(question.answers), 2)
```
Again is this not testing the add functionality therefore be another test?

samiwel replied:
I've split these out into two separate tests.

benjefferies commented on tests/app/questionnaire_state/test_state_question.py:
```diff
@@ -0,0 +1,152 @@
+
+        self.assertEqual(len(question.children), 3)
+
+        question.answers.clear()
```
Does the second part of this test below test anything different therefore needed?

samiwel replied:
I've split this out into two tests as they're two different scenarios. Second one will come into play when middle and last names are added in the next story, I was thinking ahead it seems!

benjefferies commented on tests/functional/pages/surveys/household_composition/summary.page.js:
```diff
@@ -0,0 +1,36 @@
+    return browser.element('[data-qa="414699da-1667-44fd-8e98-7606966884db-answer"]').getText() === name
+  }
+
+  areNamesDisplayed(names) {
```
It might be nicer if this was getHouseholdNames with a contains assertion instead of us writing that logic ourself

benjefferies commented on tests/functional/spec/household-composition.spec.js:
```diff
@@ -0,0 +1,114 @@
+    expect(HouseholdCompositionSummary.isNameDisplayed('Person One')).to.be.true
+  })
+
+  it('Given no people added, when enter a name and add another person, then there should be two input fields displayed.', function() {
```
Given a name has been added, when I enter another name?

samiwel replied:
I've re-worded this.

benjefferies commented on tests/functional/spec/household-composition.spec.js:
```diff
@@ -0,0 +1,114 @@
+    expect(HouseholdCompositionSummary.areNamesDisplayed(['Person One', 'Person Two'])).to.be.true
+  })
+
+  it('Given three people added, when submitted, all three names should appear on summary.', function() {
```
Maybe we don't need this test as it doesn't exercise anything more than `Given two people added, when submitted, both names should appear on summary.`

samiwel replied:
I've removed the two person one and stuck with the 3 people test.

benjefferies commented on tests/functional/spec/household-composition.spec.js:
```diff
@@ -0,0 +1,114 @@
+    expect(HouseholdCompositionSummary.areNamesDisplayed(['Person One'])).to.be.true
+  })
+
+   it('Given three people added, when I remove second person, first and third person should appear on summary.', function() {
```
Does this exercise anymore logic than `Given two people added, when I remove second person, only first person should appear on summary.` does?

samiwel replied:
It just catches the case where you remove an item from the middle. There was a bug with this during the initial spike, so I just wanted to make sure it was covered.

benjefferies replied:
That's fair enough. That sounds like it could be caught at unit test level. I am worried we might start using selenium to test logic which could be tested in a more isolated, faster way

samiwel replied:
I'll try put together an unit test to cover this functionality.

ajmaddaford commented on app/questionnaire/questionnaire_manager.py:
```diff
@@ -125,3 +126,44 @@ def get_schema_item_by_id(self, item_id):
+        new_answer.schema_item = new_answer_schema
+        new_answer.parent = question_state
+        new_answer.instance = next_instance_id
+        return new_answer
```
Can we adapt the `_add_new_answer` method in the `StateQuestion` class to create the new answer state item?

samiwel committed:
Fix tests.

ajmaddaford commented on app/questionnaire_state/state_question.py:
```diff
@@ -6,3 +8,54 @@ def __init__(self, id, schema_item):
+        new_answer.schema_item = new_answer_schema
+
+        self.children.append(new_answer)
+        return new_answer
```
If this method was modified to create a new answer state (rather than create and append) , could this method be moved to `StateAnswer` class?

benjefferies commented:
Not sure if it was intended but the focused element after adding a person is the remove person so if you tap enter to submit you remove a person

samiwel committed:
Fix and refactor code based on review comments.

samiwel committed:
Rename test file to avoid codecov conflict.

samiwel committed:
Merge branch 'master' into household-composition-repeat-single

collisdigital commented:
With respect to errors; check with Louise, but I believe there shouldn't be any; i.e. you can enter nothing and still proceed or click Add person and leave blank and that's fine. However, since this PR is huge perhaps don't change any behaviour but add a new card after discussing with Louise.

benjefferies commented on app/templating/summary/question.py:
```diff
@@ -24,13 +24,20 @@ def _build_answers(cls, block_id, question_schema, answer_schema, answers):
-            answer = cls._build_answer(question_schema, answer_schema, answers, answers_iterator)
-            summary_answers.append(Answer(block_id, answer_schema, answer))
+            if question_schema['type'] == 'RepeatingAnswer':
+                for k, v in answers.items():
```
If we don't use `v` I think the pythonic way of ignoring it is with an underscore `k, _`

samiwel committed:
Addressing PR review comments.

  - Refactor add_answer in questionnaire_manager.
  - Refactor remove_answer in questionnaire_manager.
  - Add test to cover case when schema item not found.
  - Fix repeating answers unique widget Id.
  - Refactor add_answer and remove_answer into a single endpoint.

collisdigital commented on app/data/test_household_question.json:
```diff
@@ -0,0 +1,74 @@
+            "display": {
+                "properties": {}
+            },
+            "id": "multiple_questions_group",
```
Should be multiple-questions-group with - not _

samiwel replied:
Changed

collisdigital commented on app/main/views/questionnaire.py:
```diff
@@ -119,13 +115,44 @@ def submit_answers(eq_id, form_type, collection_id):
+                             form_type=form_type, collection_id=collection_id)
+
+
+def _go_to_next_block(current_location, eq_id, form_type, collection_id, ):
```
Trailing comma, in args list.

samiwel replied:
Removed

ajmaddaford commented on app/questionnaire/questionnaire_manager.py:
```diff
@@ -125,3 +125,29 @@ def get_schema_item_by_id(self, item_id):
+        self.update_questionnaire_store(block_id)
+
+    @staticmethod
+    def _get_next_answer_instance(answer_store, existing_answer_schema):
```
We shouldn't be passing an `existing_answer_schema` object when we only need the `answer_id`. Also, I think we should move this method to the AnswerStore.

samiwel replied:
Changed this so that it passes the Id in directly.

samiwel committed:
Add integration test for household composition question.

samiwel committed:
Merge branch 'master' into household-composition-repeat-single

ajmaddaford commented on app/schema/questionnaire.py:
```diff
@@ -28,10 +28,13 @@ def add_group(self, group):
         elif item_id in self.items_by_id.keys():
             return self.items_by_id[item_id]
+        elif answer_without_instance_id in self.items_by_id.keys():
+            return self.items_by_id[answer_without_instance_id]
```
Do we make use of this anywhere? The only place I can see we pass an answer with an instance id is in the tests. The concern for me here is no item id's exist in the schema with an instance id, if we need to find an item with an instance id we should remove it before calling this method.

samiwel replied:
Removed this and stripping off the instance Id from the answer Id before this method is called.

ajmaddaford commented on app/schema/widget.py:
```diff
@@ -18,3 +17,8 @@ def get_user_input(self, post_vars):
+    @staticmethod
+    def get_id(state):
+        schema_id_with_suffix = '_'.join([state.schema_item.id, str(state.answer_instance)])
+        return state.schema_item.id if state.answer_instance == 0 else schema_id_with_suffix
```
Can we set the widget id at the same time we create the widget name?

samiwel replied:
Added Id attribute to widget. Updated each widget's render method to reference self.id instead of using the schema item id.

samiwel committed:
Fix failing unit test.

ajmaddaford committed:
Refactor after trio peer review

ajmaddaford committed:
Merge branch 'household-composition-repeat-single' of github.com:ONSdigital/eq-survey-runner into household-composition-repeat-single

ajmaddaford committed:
Fix issue with adding answer state to question state

ajmaddaford committed:
Fix sort to not use lambda

ajmaddaford commented on app/parser/v0_0_1/schema_parser.py:
```diff
@@ -74,6 +74,7 @@ def __init__(self, schema):
         self.question_factory.register_all({
             'GENERAL': GeneralQuestion,
             'DATERANGE': DateRangeQuestion,
+            'REPEATINGANSWER': GeneralQuestion,
```
We should create a specific question type class for this

ajmaddaford commented on app/templates/partials/question.html:
```diff
@@ -24,13 +24,45 @@ <h3 class="question__subtitle mars">{{question_subtitle}}</h3>
+   {%- if question.schema_item.type == 'RepeatingAnswer' %}
+     {%- for answer in question.answers %}
+       <div class="question__answer js-household-person">
+         <h3 class="neptune">Person <span class="js-household-loopindex">{{ loop.index }}</span>
```
This should be `_('Person ')`

collisdigital replied:
Even better, any way this could come from the Schema (probably one for a later PR)?

samiwel replied:
I have changed this to  `_('Person ')` for now, happy to revisit in future if necessary.

collisdigital commented on app/questionnaire_state/state_question.py:
```diff
@@ -1,8 +1,54 @@
 class StateQuestion(StateItem):
     def __init__(self, id, schema_item):
-        super().__init__(id=id, schema_item=schema_item)
+        super(StateQuestion, self).__init__(id=id, schema_item=schema_item)
```
can be `super().__init__(id=id, schema_item=schema_item)` in Python3

samiwel replied:
Changed

collisdigital commented on app/questionnaire_state/state_question.py:
```diff
@@ -1,8 +1,54 @@
+            for state_answer in self.answers:
+                state_answer.update_state(user_input)
+        else:
+            super(StateQuestion, self).update_state(user_input)
```
Can be `super().update_state(user_input)` in Python3

samiwel replied:
Changed

collisdigital commented on app/questionnaire_state/state_question.py:
```diff
@@ -1,8 +1,54 @@
+    def _iterate_over_instance_ids(cls, user_input):
+
+        answer_instance_ids = [key for key in user_input]
+        answer_instance_ids = sorted(answer_instance_ids, key=natural_order)
```
These two lines can be replaced with:

``` python
        answer_instance_ids = sorted(user_input.keys(), key=natural_order)
```

collisdigital replied:
at which stage it might be better to just pass user_input.keys() into this function and add a pydoc saying it expects to receive a list of raw answer_instance_ids; this ties up with a comment later in this review about re-using this function elsewhere.

collisdigital commented on app/schema/questionnaire.py:
```diff
@@ -28,10 +28,13 @@ def add_group(self, group):
             group.container = self
 
     def get_item_by_id(self, item_id):
+        answer_without_instance_id = item_id.split('_')[0]
```
Does this need to use the `_extract_answer_instance_id(...)` method; this simple split is likely to give false positives etc.

collisdigital commented on app/templating/summary/question.py:
```diff
@@ -24,13 +24,20 @@ def _build_answers(cls, block_id, question_schema, answer_schema, answers):
-            summary_answers.append(Answer(block_id, answer_schema, answer))
+            if question_schema['type'] == 'RepeatingAnswer':
+                for k, _ in answers.items():
+                    if k.startswith(answer_schema['id']):
```
this simple check will suffer the same problems as the original code before `_extract_answer_instance_id` was introduced; could re-use `_iterate_over_instance_ids`; if it is modified to take user_input.keys() instead of user_input then it could also take answers.keys() here as well to get the same effect?

collisdigital commented:
97 comments, that century is in sight 💯!

ajmaddaford committed:
Move create_new_answer_state to StateQuestion class

benjefferies commented on app/main/views/questionnaire.py:
```diff
@@ -125,13 +121,44 @@ def submit_answers(eq_id, form_type, collection_id):
+
+    if 'action[add_answer]' in request.form:
+        questionnaire_manager.add_answer('household-composition', 'question', answer_store)
+        return redirect_to_questionnaire_page(eq_id, form_type, collection_id, 'household-composition')
```
I think we could just return get_questionnaire(...) here to avoid the redirects as they will stay on the same url

benjefferies commented on app/main/views/questionnaire.py:
```diff
@@ -125,13 +121,44 @@ def submit_answers(eq_id, form_type, collection_id):
+    elif 'action[remove_answer]' in request.form:
+        index_to_remove = request.form.get('action[remove_answer]')
+        questionnaire_manager.remove_answer('household-composition', answer_store, index_to_remove)
+        return redirect_to_questionnaire_page(eq_id, form_type, collection_id, 'household-composition')
```
I think we could just return get_questionnaire(...) here to avoid the redirects as they will stay on the same url

samiwel committed:
Refactor code based on PR comments.

benjefferies commented on app/templating/summary/question.py:
```diff
@@ -24,13 +24,20 @@ def _build_answers(cls, block_id, question_schema, answer_schema, answers):
-            answer = cls._build_answer(question_schema, answer_schema, answers, answers_iterator)
-            summary_answers.append(Answer(block_id, answer_schema, answer))
+            if question_schema['type'] == 'RepeatingAnswer':
+                for k, _ in answers.items():
```
I guess this could be just answers.keys()

samiwel committed:
Fix codacy issue.

samiwel committed:
Refactor code base on PR comments.

samiwel committed:
Merge branch 'master' into household-composition-repeat-single

samiwel committed:
Add new question type for repeating answers.

samiwel committed:
Fix codacy issue.

samiwel committed:
Fix quantified code issue.

samiwel committed:
Fix display issue on summary screen.

samiwel committed:
Remove unnecessary redirect.

samiwel committed:
Merge branch 'master' into household-composition-repeat-single

benjefferies commented on tests/functional/pages/surveys/household_composition/summary.page.js:
```diff
@@ -0,0 +1,36 @@
+    return browser.element('[data-qa="household-full-name-answer"]').getText() === name
+  }
+
+  areNamesDisplayed(names) {
```
I think we are implementing complex logic here which is unnecessarily. We should use the chai testing framework to assert all values exist? i.e. return an array of names and assert those

benjefferies commented on tests/integration/household_composition/test_household_question.py:
```diff
@@ -0,0 +1,152 @@
+
+        return first_page
+
+    def test_regular_submit(self):
```
Can be deleted

samiwel replied:
Removed.

samiwel committed:
Refactor selenium test.

samiwel committed:
Fix codacy issues.

samiwel committed:
Remove redundant test.

samiwel committed:
Fix failing integration test.

samiwel commented:
Closing this PR, opening new PR #579 for a cleaner merge into master.
